File: AndroidFx/base/src/main/java/com/android/zp/base/view/SkinnableTextView.java
package com.android.zp.base.view;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;

import com.android.zp.base.KLog;
import com.android.zp.base.R;
import com.android.zp.base.core.ViewsMatch;
import com.android.zp.base.model.AttrsBean;


/**
 * 继承TextView兼容包，9.0源码中也是如此
 * 参考：AppCompatViewInflater.java
 * 86行 + 138行 + 206行
 */
public class SkinnableTextView extends AppCompatTextView implements ViewsMatch {

    private AttrsBean attrsBean;

    public SkinnableTextView(Context context) {
        this(context, null);
    }

    public SkinnableTextView(Context context, AttributeSet attrs) {
        this(context, attrs, android.R.attr.textViewStyle);
    }

    public SkinnableTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        attrsBean = new AttrsBean();

        // 根据自定义属性，匹配控件属性的类型集合，如：background + textColor
        TypedArray typedArray = context.obtainStyledAttributes(attrs,
                R.styleable.SkinnableTextView,
                defStyleAttr, 0);
        // 存储到临时JavaBean对象
        attrsBean.saveViewResource(typedArray, R.styleable.SkinnableTextView);
        // 这一句回收非常重要！obtainStyledAttributes()有语法提示！！
        typedArray.recycle();
    }

    @Override
    public void skinnableView() {
        // 根据自定义属性，获取styleable中的background属性
        int key = R.styleable.SkinnableTextView[R.styleable.SkinnableTextView_android_background];
        // 根据styleable获取控件某属性的resourceId
        int backgroundResourceId = attrsBean.getViewResource(key);
        KLog.logI("backgroundResourceId: " + backgroundResourceId);
        if (backgroundResourceId > 0) {
            // 兼容包转换
            Drawable drawable = ContextCompat.getDrawable(getContext(), backgroundResourceId);
            // 控件自带api，这里不用setBackgroundColor()因为在9.0测试不通过
            // setBackgroundDrawable本来过时了，但是兼容包重写了方法
            setBackgroundDrawable(drawable);
        }

        // 根据自定义属性，获取styleable中的textColor属性
        key = R.styleable.SkinnableTextView[R.styleable.SkinnableTextView_android_textColor];
        int textColorResourceId = attrsBean.getViewResource(key);
        KLog.logI("textColorResourceId: " + textColorResourceId);
        if (textColorResourceId > 0) {
            ColorStateList color = ContextCompat.getColorStateList(getContext(), textColorResourceId);
            setTextColor(color);
        }
    }
}

File: AndroidFx/base/src/main/java/com/android/zp/base/view/SkinnableTextView.kt
package com.android.zp.base.view

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.TypedArray
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import com.android.zp.base.R
import com.android.zp.base.core.ViewsMatch
import com.android.zp.base.model.AttrsBean

class SkinnableTextView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : AppCompatTextView(context, attrs, defStyleAttr), ViewsMatch {

    private var attrsBean: AttrsBean? = null

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, android.R.attr.textViewStyle)

    init {
        attrsBean = AttrsBean()
        val typedArray: TypedArray = context.obtainStyledAttributes(attrs, R.styleable.SkinnableTextView, defStyleAttr, 0)
        attrsBean?.saveViewResource(typedArray, R.styleable.SkinnableTextView)
        typedArray.recycle()
    }

    override fun skinnableView() {
        var key = R.styleable.SkinnableTextView[R.styleable.SkinnableTextView_android_background]
        val backgroundResourceId = attrsBean?.getViewResource(key)
        if (backgroundResourceId!! > 0) {
            val drawable: Drawable? = ContextCompat.getDrawable(context, backgroundResourceId)
            setBackgroundDrawable(drawable)
        }
        key = R.styleable.SkinnableTextView[R.styleable.SkinnableTextView_android_textColor]
        val textColorResourceId = attrsBean?.getViewResource(key)
        if (textColorResourceId!! > 0) {
            val color: ColorStateList? = ContextCompat.getColorStateList(context, textColorResourceId)
            setTextColor(color)
        }
    }

}